### ISSUE TITLE
Generated serialize/deserialize is wrong for ADTs

### ISSUE THREAD

**Issue by garuma** [sorbet/sorbet#3695]
#### Summary

When using Sorbet [support for approximate algebraic data types](https://sorbet.org/docs/sealed#approximating-algebraic-data-types) the serializer/deserializer that's generated doesn't correctly understand the underlying construct and generate something wrong.

Take this definition:

```ruby
# typed: true

module MessagePayload
  extend T::Helpers

  sealed!

  class MainPayload < T::Struct
    include MessagePayload

    const :foo, String
  end
end

class Message < T::Struct
  const :name, String
  const :payload, MessagePayload
end

Message.new(name: "Foo", payload: MessagePayload::MainPayload.new(foo: "bar")).serialize
```

This will output: `{"name"=>"Foo", "payload"=><MessagePayload::MainPayload foo="bar">}`

This is incorrect because the payload part should recursively have `serialize` called on it. The main issue here is that the `SerdeGenerator` doesn't have understanding that the type `MessagePayload` is a proxy for a closed subset of otherwise serializable `T::Struct` classes and thus will fallback on using `T::Props::Utils.deep_clone_object` internally.

A workaround is to apply a dummy `include T::Props::Serializable` in the module (despite the fact it would never be called in practice) so that the generator ends up correct:

``` ruby
# typed: true

module MessagePayload
  extend T::Helpers
  include T::Props::Serializable

  sealed!

  class MainPayload < T::Struct
    include MessagePayload

    const :foo, String
  end
end

class Message < T::Struct
  const :name, String
  const :payload, MessagePayload
end

Message.new(name: "Foo", payload: MessagePayload::MainPayload.new(foo: "bar")).serialize
```

This will output: `{"name"=>"Foo", "payload"=>{"foo"=>"bar"}}`

#### Expected behavior

Depending on how easy it is for the generator to understand that a module is meant to represent a discriminated union it could attempt to figure out if those all contain a serializable marker. If not, it also seems adequate that modifiers like `Serializable` be included directly on the module like with the workaround but in that case documentation should be updated to mention it.

**Comment by aprocter-stripe** [comment]
Thanks @garuma! At a bare minimum I think we'll need to document this limitation.

@jez @elliottt Thoughts on difficulty level here?

**Comment by garuma** [comment]
An extra challenge that I didn't hint at here is for deserialization since there is no discriminant that can be used by the routine currently (in my case I use a separate `T::Enum` encoded in the message and override `deserialize` to dispatch to the right sub-type in the module).

Potentially that discriminant could be something that's automatically included in the generated `serialize` call (of course out of scope for this particular bug).

PS: I personally really like Typescript technique [with tagged unions](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#discriminating-unions) to solve this problem.

**Comment by jez** [comment]
This is a big open question that we should not take quick PRs for. There is currently already a problem of serialization with `T.any` types. Arguably it's slightly worse for sealed types, but fundamentally the API of serialization for T::Props is owned by another team within Stripe (storage) and we should not make quick decisions.

**Comment by garuma** [comment]
Definitely, I would say the documentation approach is the best thing to do for now to at least mention the shortcoming. Now the question is do we think that the workaround of manually adding `include T::Props::Serializable` to the outer module is good enough for now or would that open a can of worms later?
